cmd/grpcui/grpcui.go

```diff
@@ -718,3 +725,91 @@ func logInfof(format string, args ...interface{}) {
 	prefix := "INFO: " + time.Now().Format("2006/01/02 15:04:05") + " "
 	log.Printf(prefix+format, args...)
 }
+
+func dial(ctx context.Context, network, addr string, creds credentials.TransportCredentials, failFast bool, opts ...grpc.DialOption) (*grpc.ClientConn, error) {
+	if failFast {
+		return grpcurl.BlockingDial(ctx, network, addr, creds, opts...)
+	}
+	// BlockingDial will return the first error returned. It is meant for interactive use.
+	// If we don't want to fail fast, then we need to do a more customized dial.
+
+	// TODO: perhaps this logic should be added to the grpcurl package, like in a new
+	// BlockingDialNoFailFast function?
+
+	dialer := &errTrackingDialer{
+		dialer:  &net.Dialer{},
+		network: network,
+	}
+	var errCreds errTrackingCreds
+	if creds == nil {
+		opts = append(opts, grpc.WithInsecure())
+	} else {
+		errCreds = errTrackingCreds{
+			TransportCredentials: creds,
+		}
+		opts = append(opts, grpc.WithTransportCredentials(&errCreds))
+	}
+
+	cc, err := grpc.DialContext(ctx, addr, append(opts, grpc.WithBlock(), grpc.WithContextDialer(dialer.dial))...)
+	if err == nil {
+		return cc, nil
+	}
+
+	// prefer last observed TLS handshake error if there is one
+	if err := errCreds.err(); err != nil {
+		return nil, err
+	}
+	// otherwise, use the error the dialer last error
+	if err := dialer.err(); err != nil {
+		return nil, err
+	}
+	// if we have no better source of error message, use what grpc.DialContext returned
+	return nil, err
+}
+
+type errTrackingCreds struct {
+	credentials.TransportCredentials
+
+	mu      sync.Mutex
+	lastErr error
+}
+
+func (c *errTrackingCreds) ClientHandshake(ctx context.Context, addr string, rawConn net.Conn) (net.Conn, credentials.AuthInfo, error) {
+	conn, auth, err := c.TransportCredentials.ClientHandshake(ctx, addr, rawConn)
+	if err != nil {
+		c.mu.Lock()
+		defer c.mu.Unlock()
+		c.lastErr = err
+	}
+	return conn, auth, err
+}
+
+func (c *errTrackingCreds) err() error {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+	return c.lastErr
+}
+
+type errTrackingDialer struct {
+	dialer  *net.Dialer
+	network string
+
+	mu      sync.Mutex
+	lastErr error
+}
+
+func (c *errTrackingDialer) dial(ctx context.Context, addr string) (net.Conn, error) {
+	conn, err := c.dialer.DialContext(ctx, c.network, addr)
+	if err != nil {
+		c.mu.Lock()
+		defer c.mu.Unlock()
+		c.lastErr = err
+	}
+	return conn, err
+}
+
+func (c *errTrackingDialer) err() error {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+	return c.lastErr
+}
```

Comment:
Wow... it's kind of shocking that this level of invasive wrapping is needed just to get a decent error back.

At any rate, cmd/grpcui seems like a distinctly weird place for this logic to live...  at least if there were `grpcurl.DialContextWithBetterErrors`, it could be reused.  I'll approve, but I think your TODO is prescient.